### PR TITLE
Fix nystromformer attention

### LINKED ISSUE #360: NystromAttention sometimes results in NaN values when using key_padding_mask
# 🐛 Bug

NystromAttention sometimes results in NaN values when using key_padding_mask

## To Reproduce

```python
from xformers.components.attention import NystromAttention
from xformers.components.feedforward import MLP
from xformers.components import MultiHeadDispatch
from xformers.components.attention.utils import (bool_mask_to_additive)
import torch


input_tensor = torch.load(file_path).unsqueeze(0)
input_tensor[:, -3:] = 0
print('input shape: ', input_tensor.shape)
print('input: ', input_tensor, '\n')

torch.manual_seed(0)
num_heads = 8
proj_dropout = 0.2
bias = (False, False, False, True)
attn = NystromAttention(dropout = 0., 
                num_heads = num_heads, 
                num_landmarks = 256)

msa = MultiHeadDispatch(dim_model = 1024, 
                  num_heads = num_heads, 
                  attention = attn, 
                  bias = bias,
                  residual_dropout = proj_dropout, 
                  use_separate_proj_weight = True)

padding_mask = torch.zeros(1, input_tensor.size(1), dtype=torch.bool)
padding_mask[:, -3:] = True
key_padding_mask = bool_mask_to_additive(~padding_mask)
output = msa(input_tensor, key_padding_mask = key_padding_mask)
print('output with key_padding_mask: ', output)
print()

output = msa(input_tensor, key_padding_mask = None)
print('output without key_padding_mask: ', output)
```

Printout of results:
input shape:  torch.Size([1, 24798, 1024])
input:  tensor([[[0.0864, 0.0117, 0.0049,  ..., 0.0104, 0.0113, 0.0380],
         [0.0842, 0.0025, 0.0103,  ..., 0.0043, 0.0168, 0.0368],
         [0.1197, 0.0151, 0.0195,  ..., 0.0133, 0.0086, 0.0430],
         ...,
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000]]]) 

output with key_padding_mask:  tensor([[[ 0.0269, -0.0089,  0.0000,  ..., -0.0080,  0.0022,  0.0241],
         [ 0.0269, -0.0089,  0.0468,  ..., -0.0080,  0.0022,  0.0241],
         [ 0.0269, -0.0000,  0.0468,  ..., -0.0080,  0.0022,  0.0241],
         ...,
         [    nan,     nan,     nan,  ...,     nan,     nan,     nan],
         [    nan,     nan,     nan,  ...,     nan,     nan,     nan],
         [    nan,     nan,     nan,  ...,     nan,     nan,     nan]]],
       grad_fn=\<MulBackward0\>)

output without key_padding_mask:  tensor([[[ 0.0000, -0.0089,  0.0468,  ..., -0.0080,  0.0022,  0.0000],
         [ 0.0269, -0.0089,  0.0468,  ..., -0.0080,  0.0022,  0.0241],
         [ 0.0269, -0.0000,  0.0000,  ..., -0.0000,  0.0022,  0.0241],
         ...,
         [ 0.0269, -0.0089,  0.0468,  ..., -0.0000,  0.0022,  0.0241],
         [ 0.0269, -0.0089,  0.0000,  ..., -0.0080,  0.0000,  0.0241],
         [ 0.0269, -0.0089,  0.0000,  ..., -0.0080,  0.0022,  0.0241]]],
       grad_fn=\<MulBackward0\>)

## Expected behavior
No nan values produced

## Environment

- xformers version: 0.0.11
- PyTorch Version (e.g., 1.0): 1.10.2
- OS (e.g., Linux): Linux
- How you installed PyTorch (`conda`, `pip`, source): conda
- Python version: 3.8.12
- CUDA/cuDNN version: 11.1
- GPU models and configuration: NVIDIA 3090 (i reproduced the bug on cpu as well as seen in the above code snippet where no cuda device is used).

## Additional context
Happy to share the tensor I used by email or any other means (it's a .pt file)

### PR DESCRIPTION
**Pull request by tma15**
## What does this PR do?
Fixes #360. When using `key_padding_mask`, nystroformer attention was raising NaN because of masking on calculation of `kernel_1`. Although `key_padding_mask` was used as `mask_1`, key was the compressed by pooing already.

Following the original code of [Nystromformer](https://github.com/mlpen/Nystromformer/blob/main/code/attention_nystrom.py), I have updated the corresponding codes. Concretely, padding tokens in query and key were masked before pooling, and masks of calculation in `kernel_1` and `kernel_2` were deleted.

After that, I have confirmed that `NaN` was not raised. In addition, I have deleted replacement of NaN with random values in the test because existing NaN in output are not expected.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Review by dianaml0 — APPROVED** [review]
Thanks a lot for this fix!
